src/xc_integrator/replicated/device/incore_replicated_xc_device_integrator_exc_vxc.hpp

```diff
@@ -225,95 +412,84 @@ void IncoreReplicatedXCDeviceIntegrator<ValueType>::
       device_data.generate_buffers( enabled_terms, basis_map, task_it, task_end );
 
     /*** Process the batches ***/
-
+    
     // Evaluate collocation
     if( func.is_gga() ) lwd->eval_collocation_gradient( &device_data );
     else                lwd->eval_collocation( &device_data );
-
-    // Evaluate X matrix
-    lwd->eval_xmat( 2.0, &device_data );
-
-    // Evaluate U/V variables
-    if( func.is_gga() ) lwd->eval_uvvar_gga_rks( &device_data );
-    else                lwd->eval_uvvar_lda_rks( &device_data );
+    
+    auto* data = dynamic_cast<XCDeviceAoSData*>(&device_data);
+    auto tasks = data->host_device_tasks;
+    auto& task = tasks[0];
+    auto static_stack = data->static_stack;
+    auto base_stack   = data->base_stack;
+    
+    const double xmat_fac = is_rks ? 2.0 : 1.0;
+
+    // Evaluate X matrix and V vars
```

Comment:
This could be cleaned up with a lambda
```
  auto do_xmat_vvar = [&](den_id) {
    lwd->eval_xmat( xmat_fac, &device_data, do_xgrad, den_id );
    lwd->eval_vvar( &device_data, func.is_gga(), den_id );
  };
  
  do_xmat_vvar(DEN_S);
  if(not is_rks) {
    do_xmat_vvar(DEN_Z);
    if(not is_uks) {
      do_xmat_vvar(DEN_Y);
      do_xmat_vvar(DEN_X);
    }
  }
```